### PR TITLE
230 notify of season switch when opening inspection

### PR DESCRIPTION
**Pull request by danieldunderfelt**
Update #75 to show a notification when the season changes.

Issue: https://github.com/HSLdevcom/bultti/issues/230

Changes:

- Create new InfoMessages component that can show info and error messages.
- Show info message when the season changes automatically
- Show messages as toasts
- Refactor state and related utilities for more type safety and compatibility

### CODE REVIEW

File: src/util/useShowNotification.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { useStateValue } from '../state/useAppState'
+import { useCallback } from 'react'
+import { UINotification } from '../type/state'
+
+export function useShowNotification(type: UINotification['type']) {
+  let [, { add }] = useStateValue('notifications')
+  return useCallback((message: string) => add({ message, type }), [add])
+}
```
Comment (Divinare) [inline src/util/useShowNotification.ts:8]: Now we have diff naming:
var setInfoMessage = useShowInfoNotification()
let showErrorNotification = useShowErrorNotification()
var setErrorMessage = useShowErrorNotification()
let showErrorMessage = useShowErrorNotification()

These should be renamed as:
var showInfoNotification = useShowInfoNotification()
let showErrorNotification = useShowErrorNotification()
var showErrorNotification = useShowErrorNotification()
let showErrorNotification = useShowErrorNotification()

But idk, maybe i'd prefer this style:
```
type MessageType = 'info' | 'error'
export const showNotification = ({ message, type = 'info' }: { message: string, type?: MessageType }) => {...}
showNotification({ message: '...', type: 'error'})
```
At least this with approach you don't end up using different names for the function :P which again slightly improves readability. But yea I also agree one can see this as "I prefer this car" type of things

File: src/text/fi.json [hunk 2]
```diff
@@ -169,6 +172,7 @@
   "inspectionPage_sanctions": "Sanktiointi",
   "inspectionPage_relatedUsers": "Käyttäjät",
 
+  "inspection_seasonChangedAutomatically": "The season was automatically changed to ${newSeason}",
```
Comment (Divinare) [inline src/text/fi.json:175]: Finnish?

**Review by Divinare — CHANGES_REQUESTED** [review]
Testing:

Styling looks greatly better now!

I noticed that if you have lets say an existing notification "The season was automatically changed to SY2020", it wont show up again if it's switched back. I would show duplicate messages OR you could flash the existing notification to indicate that it occured again (the latter is what I did in a previous project where it worked well). Maybe here i'd go with the first option however because errors / messages are shown quite infrequently.

Otherwise this seems to work well, a nice addition were those globalState fixes + that there is no longer an empty option in global selectors

File: src/text/fi.json (hunk 2 above)
Comment (danieldunderfelt) [reply src/text/fi.json:175]: Hahah yes... :D

File: src/util/useShowNotification.ts (hunk 1 above)
Comment (danieldunderfelt) [reply src/util/useShowNotification.ts:8]: Good point about naming, I'll unify the names. When refactoring I just aliased `add` to the name already used in the code to show an error message, which was mostly `setErrorMessage`. This resulted in less errors and less changes, but maybe it's better to use the same name everywhere.

For the hook itself, I like using different functions for showing different types, but the general `useShowNotification` can also be used.

**Comment by danieldunderfelt** [comment]
> I noticed that if you have lets say an existing notification "The season was automatically changed to SY2020", it wont show up again if it's switched back. I would show duplicate messages OR you could flash the existing notification to indicate that it occured again (the latter is what I did in a previous project where it worked well but there the same error happened quite often). Maybe here i'd go with the first option however because errors / messages are shown quite infrequently.

Yes, we should add some nice animations and your suggestions are good, but I wanted to keep it simple for now as this task got way too big already. We can continue working on the notifications in another issue.

**Comment by danieldunderfelt** [comment]
> I noticed that if you have lets say an existing notification "The season was automatically changed to SY2020", it wont show up again if it's switched back. I would show duplicate messages OR you could flash the existing notification to indicate that it occured again (the latter is what I did in a previous project where it worked well but there the same error happened quite often). Maybe here i'd go with the first option however because errors / messages are shown quite infrequently.

Yes, we should add some nice animations and your suggestions are good, but I wanted to keep it simple for now as this task got way too big already. We can continue working on the notifications in another issue.

**Comment by Divinare** [comment]
Created a card for handling duplicated notifications: https://github.com/HSLdevcom/bultti/issues/269

**Pull request merged**